bootstrapvz/common/tasks/filesystem.py

```diff
@@ -100,17 +100,73 @@ def run(cls, info):
 
 
 class MountSpecials(Task):
-	description = 'Mounting special block devices'
+	description = 'Mounting special devices'
 	phase = phases.os_installation
 	predecessors = [bootstrap.Bootstrap]
 
 	@classmethod
 	def run(cls, info):
 		root = info.volume.partition_map.root
-		root.add_mount('/dev', 'dev', ['--bind'])
+
+		# Create /dev
+		# See https://git.kernel.org/cgit/linux/kernel/git/torvalds/linux.git/tree/Documentation/devices.txt
+		import os, stat
+		from os import mknod, makedev, symlink
+		from os.path import join
+
+                # First, make a tmpfs out of /dev.  Everything must be gone after umount.
+		root.add_mount('none', 'dev', ['--types', 'tmpfs'])
+		dev = join(info.root, 'dev')
+
+                # The GNU libc expects /dev/shm to be a world-writable directory in a tmpfs
+                # This is used by the POSIX shared memory implementation
+		os.makedirs(join(dev, 'shm'), 0o777)
+
+                # The API pseudo-devices (null, zero, full, random, urandom and tty)
+		for name, major, minor in [ ('null',   1, 3), ('zero',    1, 5), ('full', 1, 7),
+		                            ('random', 1, 8), ('urandom', 1, 9), ('tty',  5, 0) ]:
+			mknod(join(dev, name), 0o666 | stat.S_IFCHR, makedev(major, minor))
+
+                # The nbd devices
+                os.makedirs(join(dev, 'mapper'), 0o755)
+                nbd_max_part = int(info.volume._module_param('nbd', 'max_part'))
+                for i in xrange(32 / nbd_max_part):
+                        mknod(join(dev, 'nbd%i' % i),
+                              0o660 | stat.S_IFBLK,
+                              makedev(43, nbd_max_part * i))
+                        for j in xrange(nbd_max_part):
+                                mknod(join(dev, 'nbd%ip%i' % (i,j+1)),
+                                      0o660 | stat.S_IFBLK,
+                                      makedev(43, nbd_max_part * i + j+1))
+
+                                path = '/dev/mapper/nbd%ip%i' % (i,j+1)
+                                if os.path.exists(path):
```

Comment:
Under which circumstances would this path not be a link? Shouldn't we consider that an error? We expect something to be there, but it's not, which means somethings fishy.